src/net/ossl.cc

```diff
@@ -1170,30 +1189,34 @@ class session : public enable_shared_from_this<session>, public session_impl {
         SSL_CTX_set1_cert_store(ssl_ctx.get(), *_creds);
 
         if (_creds->get_priority_string() != "") {
-            if (SSL_CTX_set_cipher_list(ssl_ctx.get(), _creds->get_priority_string().c_str()) != 1) {
-                throw ossl_error("Failed to set priority list");
-            }
+            // if (SSL_CTX_set_cipher_list(ssl_ctx.get(), _creds->get_priority_string().c_str()) != 1) {
+            //     throw ossl_error("Failed to set priority list");
+            // }
         }
         return ssl_ctx;
     }
 
-    static std::optional<sstring> get_ossl_string(X509_NAME* name){
-        if (auto name_str = X509_NAME_oneline(name, nullptr, 0)) {
-            // sstring constructor may throw, to ensure deallocation of this OpenSSL string in
-            // all cases, wrap the call to free() in a deferred_action
-            auto done = defer([&name_str]() noexcept { OPENSSL_free(name_str); });
-            sstring ossl_str(name_str);
-            return ossl_str;
+    static std::optional<sstring> get_dn_string(X509_NAME* name) {
+        auto out = bio_ptr(BIO_new(BIO_s_mem()));
+        if (-1 == X509_NAME_print_ex(out.get(), name, 0, ASN1_STRFLGS_RFC2253 | XN_FLAG_SEP_COMMA_PLUS |
+                                     XN_FLAG_FN_SN | XN_FLAG_DUMP_UNKNOWN_FIELDS)) {
+            return std::nullopt;
         }
-        return std::nullopt;
+        char * bio_ptr = nullptr;
+        auto len = BIO_get_mem_data(out.get(), &bio_ptr);
+        if (len < 0) {
+            throw ossl_error("Failed to allocate DN string");
+        }
+        return sstring(bio_ptr, len);
     }
 
     future<> client_handshake() {
         return do_handshake(SSL_connect, [this]{
             return pull_encrypted_and_send().then([this]{
                 return wait_for_input().then([this]{
                     if (eof()) {
-                        return make_exception_future<>(std::runtime_error("EOF observed during handshake"));
+                        _error = std::make_exception_ptr(std::system_error(ENOTCONN, std::system_category()));
```

Comment:
Is it possible to add a test?

The commit message talks about Redpanda, but I think it would be more appropriate to discuss this in terms of compatibility with existing behaviour in GnuTLS.

Reply:
> The commit message talks about Redpanda, but I think it would be more appropriate to discuss this in terms of compatibility with existing behaviour in GnuTLS.

In this situation, GnuTLS will return an error code which turns into a `system_error` which Redpanda does parse to see if it's a reconnect type error.